src/wallet/wallet.cpp

```diff
@@ -3523,18 +3576,78 @@ void CReserveKey::ReturnKey()
     vchPubKey = CPubKey();
 }
 
+void CWallet::CheckKeypoolMinSize() {
+    unsigned int keypool_min = GetArg("-keypoolmin", DEFAULT_KEYPOOL_MIN);
+    if (IsHDEnabled() && (setInternalKeyPool.size() < keypool_min || (setExternalKeyPool.size() < keypool_min))) {
+        // if the remaining keypool size is below the gap limit, shutdown
+        LogPrintf("%s: Keypool is too small. Shutting down. internal keypool: %d, external keypool: %d, keypool minimum: %d\n",
+                  __func__, setInternalKeyPool.size(), setExternalKeyPool.size(), keypool_min);
+        const static std::string error_msg = "Keypool is too small. Shutting down";
+        uiInterface.ThreadSafeMessageBox(error_msg, "", CClientUIInterface::MSG_ERROR);
+        StartShutdown();
+        throw std::runtime_error(error_msg);
+    }
+}
+
 static void LoadReserveKeysToSet(std::set<CKeyID>& setAddress, const std::set<int64_t>& setKeyPool, CWalletDB& walletdb) {
     for (const int64_t& id : setKeyPool)
     {
         CKeyPool keypool;
         if (!walletdb.ReadPool(id, keypool))
-            throw std::runtime_error(std::string(__func__) + ": read failed");
+            throw std::runtime_error(strprintf("%s: read failed for index %d", __func__, id));
         assert(keypool.vchPubKey.IsValid());
         CKeyID keyID = keypool.vchPubKey.GetID();
         setAddress.insert(keyID);
     }
 }
 
+void CWallet::MarkReserveKeysAsUsed(const CKeyID& keyId)
+{
+    AssertLockHeld(cs_wallet);
+    CWalletDB walletdb(*dbw);
+    for (std::set<int64_t> *setKeyPool : {&setInternalKeyPool, &setExternalKeyPool}) {
+        int64_t foundIndex = -1;
+        for (const int64_t& id : *setKeyPool) {
```

Comment:
Maybe use `index` instead of `id` here and other places to distinguish from pool indices from key ids.